stylus-sdk/src/storage/bytes.rs

```diff
@@ -231,33 +221,173 @@ impl StorageBytes {
                 len -= 32;
             }
         }
-        unsafe { Storage::clear_word(self.root) };
+        // set length and data in root storage to zero
+        unsafe { Storage::clear_word(root.storage.root) };
     }
 }
 
-// TODO: efficient bulk insertion
 impl Extend<u8> for StorageBytes {
     fn extend<T: IntoIterator<Item = u8>>(&mut self, iter: T) {
-        for elem in iter {
-            self.push(elem);
+        let mut root = BytesRoot::new_mut(self);
+        let old_len = root.len();
+        let mut chunk = Vec::with_capacity(32);
+        let mut iter = iter.into_iter();
+        let mut len = old_len;
+
+        // get the current tail to begin writing. unused if resulting data length is < 32
+        let mut slot = if old_len < 32 {
+            *root.storage.base()
+        } else {
+            let (slot, _) = root.index_slot(old_len - 1);
+            slot
+        };
+
+        // we want to work with word-aligned chunks, fill in first chunk to get there
+        if old_len % 32 != 0 {
+            if old_len < 32 {
+                // short data, get from root word
+                chunk.extend_from_slice(&root.word[..old_len]);
+            } else {
+                // long data, get from last word
+                let (slot, offset) = root.index_slot(old_len - 1);
+                let stored_word = Storage::get_word(slot);
+                chunk.extend_from_slice(&stored_word[..offset as usize + 1]);
+            }
+            len -= chunk.len(); // this will be added back as an entire word
+
+            while chunk.len() < 32 {
+                if let Some(byte) = iter.next() {
+                    chunk.push(byte);
+                } else {
+                    break;
+                }
+            }
+
+            // if total data length < 32, store short represenatation
+            if chunk.len() < 32 {
+                root.word[old_len..chunk.len()].copy_from_slice(&chunk[old_len..]);
+                unsafe {
+                    root.write_len(chunk.len());
+                }
+                return;
+            }
+
+            // write the word we just filled in
+            unsafe {
+                Storage::set_word(slot, B256::from_slice(&chunk));
+            }
+            chunk.clear();
+            len += 32;
+            slot += U256::from(1);
+        }
+
+        // write to storage, a word at a time
+        for byte in iter {
+            chunk.push(byte);
+            if chunk.len() == 32 {
+                unsafe {
+                    Storage::set_word(slot, B256::from_slice(&chunk));
+                }
+                chunk.clear();
+                len += 32;
+                slot += U256::from(1);
+            }
+        }
+
+        if !chunk.is_empty() {
+            unsafe {
+                Storage::set_word(slot, B256::right_padding_from(&chunk));
+            }
+            len += chunk.len();
+        }
+        unsafe {
+            root.write_len(len);
```

Comment:
There might be a bug when old length equals zero and the chunk is not empty. The code will store to the root storage directly with `Storage::set_word`, and then the `root.write_len` will overwrite the root storage when setting the length.

(This logic is complicated, so I might be wrong.)